doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,879 @@
+`versin` remains unevaluated.
+
+```{note}
+
+`Function` subclasses should not redefine `__new__` or `__init__`. If you want
+to implement behavior that isn't possible with `eval`, it might make more
+sense to subclass {class}`~.Expr` rather than `Function`.
+
+```
+
+`eval` can take any number of arguments, including an arbitrary number with
+`*args` and optional keyword arguments. The `.args` of the function will
+always be the arguments that were passed in by the user. For example
+
+```py
+>>> class f(Function):
+...     @classmethod
+...     def eval(cls, x, y=1, *args):
+...         return None
+```
+
+```py
+>>> f(1).args
+(1,)
+>>> f(1, 2).args
+(1, 2)
+>>> f(1, 2, 3).args
+(1, 2, 3)
+```
+
+Finally, note that automatic evaluation on floating-point inputs happens
+automatically once `evalf` is defined. See the section on [defining numerical
+evaluation with `evalf`](custom-functions-evalf) below.
+
+(custom-functions-eval-best-practices)=
+#### Best Practices for `eval`
+
+- **Don't just return an expression.**
+
+  In the above example, we might have been tempted to write
+
+  ```py
+  >>> from sympy import cos
+  >>> class versin(Function):
+  ...    @classmethod
+  ...    def eval(cls, x):
+  ...        return 1 - cos(x)
+  ```
+
+  However, this would make it so that `versin(x)` would *always* return `1 -
+  cos(x)`, regardless of what `x` is. If all you want is a quick shorthand to
+  `1 - cos(x)`, that is fine, but would be much simpler and more explicit to
+  just [use a Python function as described
+  above](custom-functions-fully-evaluated). If we defined `versin` like this,
+  it would never actually be represented as `versin(x)`, and none of the other
+  behavior we define below would matter, because the other behaviors we are
+  going to define on the `versin` class only apply when the returned object is
+  actually a `versin` instance. So for example, `versin(x).diff(x)` would
+  actually just be `(1 - cos(x)).diff(x)`, instead of calling [the `fdiff`
+  method we define below](custom-functions-differentiation).
+
+  **The point of `eval` is not to define what the function *is*,
+  mathematically, but rather to specify on what inputs it should automatically
+  evaluate.** The mathematical definition of a function is determined through
+  specifying various mathematical properties with the methods outlined below,
+  like [numerical evaluation](custom-functions-evalf),
+  [differentiation](custom-functions-differentiation), and so on.
+
+  If you find yourself doing this, you should think about what you actually
+  want to achieve. If you just want a shorthand function for an expression, it
+  will be simpler to just [define a Python
+  function](custom-functions-fully-evaluated). If you really do want a
+  symbolic function, think about when you want it to evaluate to something
+  else and when you want it to stay unevaluated.
+
+- **Avoid too much automatic evaluation.**
```

Comment:
Per [SymPy tone preferences](https://docs.sympy.org/dev/guides/contributing/documentation-style-guide.html#tone-preferences), suggest phrasing this positively. For example, "Use an appropriate amount of  automatic evaluation".

Reply:
Are you objecting to the use of the word "avoid"?